spyder/api/plugins.py

```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+        if self._PLUGINS is None:
+            self._PLUGINS = {}
+
+        if self.NAME is None:
+            raise Exception('A Spyder Plugin must define a `NAME`!')
+
+        if self.NAME in self._PLUGINS:
+            raise Exception('A Spyder Plugin with NAME="{}" already exists!'
+                            ''.format(self.NAME))
+
+        # Setup configuration
+        # --------------------------------------------------------------------
+        self._PLUGINS[self.NAME] = self
+        if self._conf is not None:
+            self._conf.register_plugin(self)
+
+        # Signals
+        # --------------------------------------------------------------------
+        self._widget.sig_option_changed.connect(self.sig_option_changed)
+        self.sig_option_changed.connect(self.set_option)
+        self._is_registered = True
+
+    def _unregister(self):
+        """
+        Disconnect signals and clean up the plugin to be able to stop a plugin
+        while Spyder is running.
+        """
+        self.sig_option_changed.disconnect()
+        self._conf.unregister_plugin()
+        self._is_registered = False
+        self._is_compatible = None
+        self._widget = None
+
+    def _register_shortcut(self, qaction, context, name,
+                           add_shortcut_to_tip=False):
+        """
+        Register a shortcut associated to a QAction or a QShortcut to
+        Spyder main application.
+
+        Parameters
+        ----------
+        qaction: QAction
+            QAction to register the shortcut for.
+        context: str
+            Name of the plugin this shortcut applies to. For instance,
+            if you pass 'Editor' as context, the shortcut will only
+            work when the editor is focused.
+            Note: You can use '_' if you want the shortcut to be work
+            for the entire application.
+        name: str
+            Name of the action the shortcut refers to (e.g. 'Debug
+            exit').  TODO: This could be overriden by the actual
+            action `name`(not the text!)
+        add_shortcut_to_tip: bool
+            If True, the shortcut is added to the action's tooltip.
+            This is useful if the action is added to a toolbar and
+            users hover it to see what it does.
+        """
+        self.sig_shortcut_registration_requested.emit(
+            qaction,
+            context,
+            name,
+            add_shortcut_to_tip,
+        )
+
+    # --- API: available methods ---------------------------------------------
+    # ------------------------------------------------------------------------
+    def create_action(self, name, text, icon=None, tip=None, toggled=None,
+                      triggered=None, data=None, menu_role=None,
+                      context=Qt.WindowShortcut, shortcut_context=None,
+                      expose=True, disable_if_not_visible=True, parent=None):
```

Comment:
>This function does not actually set a shortcut, right? How is a plugin supposed to do that?

There is no need to set shortcuts right now. We only create actions with this (and similar methods) and these are then exposed automagically as possible shortcuts. If a shortcut is found in the default config then it is assigned, otherwise is left blank for the user to use.